File: shared/internal-auth.js
export const INTERNAL_AUTH_HEADER = "x-wdl-internal-auth";
export const INTERNAL_AUTH_ENV = "WDL_INTERNAL_AUTH_TOKEN";
export const INTERNAL_AUTH_PREVIOUS_ENV = "WDL_INTERNAL_AUTH_PREVIOUS_TOKEN";
export const INTERNAL_AUTH_FAILURE_CODE = "internal_auth_failed";
export const INTERNAL_AUTH_FAILURE_MESSAGE = "Internal authentication failed";

/**
 * @param {string} name
 * @param {string} token
 */
function assertHeaderToken(name, token) {
  if (token === "") {
    throw new Error(`${name} must be configured as visible ASCII without whitespace or commas`);
  }
  for (let i = 0; i < token.length; i++) {
    const code = token.charCodeAt(i);
    if (code < 0x21 || code > 0x7e || code === 0x2c) {
      throw new Error(`${name} must be configured as visible ASCII without whitespace or commas`);
    }
  }
  return token;
}

/**
 * @param {Record<string, unknown> | null | undefined} env
 */
export function internalAuthToken(env) {
  const token = env?.[INTERNAL_AUTH_ENV];
  if (typeof token !== "string" || token === "") {
    throw new Error(`${INTERNAL_AUTH_ENV} must be configured`);
  }
  return assertHeaderToken(INTERNAL_AUTH_ENV, token);
}

/**
 * @param {Record<string, unknown> | null | undefined} env
 */
export function internalAuthPreviousToken(env) {
  const token = env?.[INTERNAL_AUTH_PREVIOUS_ENV];
  return typeof token === "string" && token !== "" ? assertHeaderToken(INTERNAL_AUTH_PREVIOUS_ENV, token) : null;
}

/**
 * @param {string} actual
 * @param {string} expected
 */
function constantTimeEqual(actual, expected) {
  const max = Math.max(actual.length, expected.length);
  let diff = actual.length ^ expected.length;
  for (let i = 0; i < max; i++) {
    diff |= (actual.charCodeAt(i) || 0) ^ (expected.charCodeAt(i) || 0);
  }
  return diff === 0;
}

/**
 * @param {Headers} headers
 * @param {Record<string, unknown> | null | undefined} env
 */
export function verifyInternalAuthHeaders(headers, env) {
  try {
    const expected = internalAuthToken(env);
    const actual = headers.get(INTERNAL_AUTH_HEADER) || "";
    if (actual === "") return false;
    if (constantTimeEqual(actual, expected)) return true;
    const previous = internalAuthPreviousToken(env);
    return previous !== null && constantTimeEqual(actual, previous);
  } catch {
    return false;
  }
}

/**
 * @param {HeadersInit | undefined} headers
 * @param {Record<string, unknown> | null | undefined} env
 */
export function withInternalAuth(headers, env) {
  const out = new Headers(headers);
  out.set(INTERNAL_AUTH_HEADER, internalAuthToken(env));
  return out;
}

/**
 * Build authenticated headers without calling `Headers#set` with the token.
 * Use this for backend capabilities callable from generated tenant facades,
 * where tenant code may have monkeypatched `Headers.prototype.set`.
 *
 * @param {HeadersInit | undefined} headers
 * @param {Record<string, unknown> | null | undefined} env
 * @returns {Record<string, string>}
 */
export function withInternalAuthEntries(headers, env) {
  /** @type {Record<string, string>} */
  const out = /** @type {Record<string, string>} */ (/** @type {unknown} */ ({ __proto__: null }));
  if (headers) {
    if (typeof /** @type {{ [Symbol.iterator]?: unknown }} */ (headers)[Symbol.iterator] === "function") {
      for (const entry of /** @type {Iterable<[unknown, unknown]>} */ (headers)) {
        const [name, value] = entry;
        if (String(name).toLowerCase() !== INTERNAL_AUTH_HEADER) {
          out[String(name)] = String(value);
        }
      }
    } else {
      for (const [name, value] of Object.entries(headers)) {
        if (name.toLowerCase() !== INTERNAL_AUTH_HEADER) {
          out[name] = String(value);
        }
      }
    }
  }
  out[INTERNAL_AUTH_HEADER] = internalAuthToken(env);
  return out;
}

/** @param {Headers} headers */
export function stripInternalAuthHeader(headers) {
  headers.delete(INTERNAL_AUTH_HEADER);
  return headers;
}

export function internalAuthFailureResponse() {
  return Response.json({
    error: INTERNAL_AUTH_FAILURE_CODE,
    message: INTERNAL_AUTH_FAILURE_MESSAGE,
  }, { status: 401 });
}

File: rust/common/src/internal_auth.rs
use std::env;

#[cfg(feature = "axum")]
use axum::{
    http::{HeaderMap, StatusCode, header::CONTENT_TYPE},
    response::{IntoResponse, Response},
};
use subtle::ConstantTimeEq;

pub const INTERNAL_AUTH_HEADER: &str = "x-wdl-internal-auth";
pub const INTERNAL_AUTH_ENV: &str = "WDL_INTERNAL_AUTH_TOKEN";
pub const INTERNAL_AUTH_PREVIOUS_ENV: &str = "WDL_INTERNAL_AUTH_PREVIOUS_TOKEN";
pub const INTERNAL_AUTH_FAILURE_CODE: &str = "internal_auth_failed";
pub const INTERNAL_AUTH_FAILURE_MESSAGE: &str = "Internal authentication failed";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct InternalAuthTokens {
    pub current: String,
    pub previous: Option<String>,
}

fn validate_internal_auth_token(name: &str, token: String) -> Result<String, String> {
    if token.is_empty()
        || token
            .bytes()
            .any(|byte| !(0x21..=0x7e).contains(&byte) || byte == b',')
    {
        return Err(format!(
            "{name} must be configured as visible ASCII without whitespace or commas"
        ));
    }
    Ok(token)
}

pub fn internal_auth_token_from_env() -> Result<String, String> {
    let token = env::var(INTERNAL_AUTH_ENV)
        .map_err(|_| format!("{INTERNAL_AUTH_ENV} must be configured"))?;
    validate_internal_auth_token(INTERNAL_AUTH_ENV, token)
}

pub fn internal_auth_tokens_from_env() -> Result<InternalAuthTokens, String> {
    let current = internal_auth_token_from_env()?;
    let previous = match env::var(INTERNAL_AUTH_PREVIOUS_ENV) {
        Ok(token) if token.is_empty() => None,
        Ok(token) => Some(validate_internal_auth_token(
            INTERNAL_AUTH_PREVIOUS_ENV,
            token,
        )?),
        Err(_) => None,
    };
    Ok(InternalAuthTokens { current, previous })
}

pub fn internal_auth_matches(actual: Option<&str>, expected: &str) -> bool {
    let Some(actual) = actual else {
        return false;
    };
    if actual.is_empty() {
        return false;
    }
    let actual = actual.as_bytes();
    let expected = expected.as_bytes();
    let max = actual.len().max(expected.len());
    let mut equal = actual.len().ct_eq(&expected.len());
    for i in 0..max {
        let actual_byte = *actual.get(i).unwrap_or(&0);
        let expected_byte = *expected.get(i).unwrap_or(&0);
        equal &= actual_byte.ct_eq(&expected_byte);
    }
    bool::from(equal)
}

pub fn internal_auth_matches_any(actual: Option<&str>, expected: &InternalAuthTokens) -> bool {
    internal_auth_matches(actual, expected.current.as_str())
        || expected
            .previous
            .as_deref()
            .is_some_and(|previous| internal_auth_matches(actual, previous))
}

#[cfg(feature = "axum")]
pub fn internal_auth_header_value(headers: &HeaderMap) -> Option<&str> {
    let mut values = headers.get_all(INTERNAL_AUTH_HEADER).iter();
    let value = values.next()?.to_str().ok()?;
    if values.next().is_some() {
        return None;
    }
    Some(value)
}

#[cfg(feature = "axum")]
pub fn internal_auth_headers_match(headers: &HeaderMap, expected: &InternalAuthTokens) -> bool {
    internal_auth_matches_any(internal_auth_header_value(headers), expected)
}

#[cfg(feature = "axum")]
pub fn internal_auth_failure_response() -> Response {
    let body = serde_json::json!({
        "error": INTERNAL_AUTH_FAILURE_CODE,
        "message": INTERNAL_AUTH_FAILURE_MESSAGE,
    })
    .to_string();
    (
        StatusCode::UNAUTHORIZED,
        [(CONTENT_TYPE, "application/json")],
        body,
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn contract() -> serde_json::Value {
        serde_json::from_str(include_str!(
            "../../../tests/fixtures/internal-auth-contract.json"
        ))
        .expect("internal auth contract fixture")
    }

    #[test]
    fn internal_auth_literals_and_rotation_match_contract() {
        let contract = contract();
        assert_eq!(contract["header"], INTERNAL_AUTH_HEADER);
        assert_eq!(contract["currentEnv"], INTERNAL_AUTH_ENV);
        assert_eq!(contract["previousEnv"], INTERNAL_AUTH_PREVIOUS_ENV);
        assert_eq!(contract["failure"]["error"], INTERNAL_AUTH_FAILURE_CODE);
        assert_eq!(
            contract["failure"]["message"],
            INTERNAL_AUTH_FAILURE_MESSAGE
        );

        for case in contract["tokenCases"].as_array().expect("tokenCases array") {
            let value = case["value"].as_str().expect("token value").to_string();
            assert_eq!(
                validate_internal_auth_token(INTERNAL_AUTH_ENV, value).is_ok(),
                case["accepted"].as_bool().expect("accepted boolean")
            );
        }

        for case in contract["rotationCases"]
            .as_array()
            .expect("rotationCases array")
        {
            let tokens = InternalAuthTokens {
                current: case["current"].as_str().expect("current token").to_string(),
                previous: case["previous"].as_str().map(str::to_string),
            };
            assert_eq!(
                internal_auth_matches_any(case["actual"].as_str(), &tokens),
                case["accepted"].as_bool().expect("accepted boolean")
            );
        }

        #[cfg(feature = "axum")]
        for case in contract["headerCases"]
            .as_array()
            .expect("headerCases array")
        {
            let tokens = InternalAuthTokens {
                current: case["current"].as_str().expect("current token").to_string(),
                previous: case["previous"].as_str().map(str::to_string),
            };
            let mut headers = HeaderMap::new();
            for value in case["values"].as_array().expect("header values") {
                headers.append(
                    INTERNAL_AUTH_HEADER,
                    value.as_str().expect("header value").parse().unwrap(),
                );
            }
            assert_eq!(
                internal_auth_headers_match(&headers, &tokens),
                case["accepted"].as_bool().expect("accepted boolean")
            );
        }
    }

    #[test]
    fn internal_auth_matches_requires_exact_token() {
        assert!(internal_auth_matches(Some("secret"), "secret"));
        assert!(!internal_auth_matches(Some("secret"), "secreu"));
        assert!(!internal_auth_matches(Some("secre"), "secret"));
        assert!(!internal_auth_matches(Some("secret"), "secret2"));
        assert!(!internal_auth_matches(Some(""), "secret"));
        assert!(!internal_auth_matches(None, "secret"));
    }

    #[test]
    fn internal_auth_matches_any_accepts_current_or_previous() {
        let tokens = InternalAuthTokens {
            current: "new".to_string(),
            previous: Some("old".to_string()),
        };
        assert!(internal_auth_matches_any(Some("new"), &tokens));
        assert!(internal_auth_matches_any(Some("old"), &tokens));
        assert!(!internal_auth_matches_any(Some("wrong"), &tokens));
        assert!(!internal_auth_matches_any(None, &tokens));
    }

    #[cfg(feature = "axum")]
    #[test]
    fn internal_auth_headers_match_accepts_current_or_previous_header() {
        let tokens = InternalAuthTokens {
            current: "new".to_string(),
            previous: Some("old".to_string()),
        };
        let mut headers = HeaderMap::new();

        headers.insert(INTERNAL_AUTH_HEADER, "new".parse().unwrap());
        assert!(internal_auth_headers_match(&headers, &tokens));

        headers.insert(INTERNAL_AUTH_HEADER, "old".parse().unwrap());
        assert!(internal_auth_headers_match(&headers, &tokens));

        headers.insert(INTERNAL_AUTH_HEADER, "wrong".parse().unwrap());
        assert!(!internal_auth_headers_match(&headers, &tokens));
    }

    #[test]
    fn internal_auth_tokens_must_be_visible_ascii_without_commas() {
        assert_eq!(
            validate_internal_auth_token(INTERNAL_AUTH_ENV, "ascii-token".to_string()).unwrap(),
            "ascii-token"
        );
        assert!(validate_internal_auth_token(INTERNAL_AUTH_ENV, "".to_string()).is_err());
        assert!(validate_internal_auth_token(INTERNAL_AUTH_ENV, "tokén".to_string()).is_err());
        assert!(
            validate_internal_auth_token(INTERNAL_AUTH_ENV, "token,other".to_string()).is_err()
        );
        assert!(validate_internal_auth_token(INTERNAL_AUTH_ENV, " token".to_string()).is_err());
        assert!(validate_internal_auth_token(INTERNAL_AUTH_ENV, "token ".to_string()).is_err());
        assert!(
            validate_internal_auth_token(INTERNAL_AUTH_ENV, "token\0value".to_string()).is_err()
        );
    }

    #[cfg(feature = "axum")]
    #[tokio::test]
    async fn internal_auth_failure_response_matches_contract() {
        let contract = contract();
        let response = internal_auth_failure_response();
        assert_eq!(
            u64::from(response.status().as_u16()),
            contract["failure"]["status"]
                .as_u64()
                .expect("failure status")
        );
        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .expect("internal auth failure body");
        let body: serde_json::Value =
            serde_json::from_slice(&body).expect("internal auth failure JSON");
        assert_eq!(body["error"], contract["failure"]["error"]);
        assert_eq!(body["message"], contract["failure"]["message"]);
    }
}
